pkg/reconciler/timeout_handler_test.go

```diff
@@ -259,3 +259,108 @@ func TestWithNoFunc(t *testing.T) {
+		},
+		{
+			description:      "returns an end time using the default timeout if a TaskRun's timeout is nil",
+			inputDuration:    nil,
+			expectedDuration: defaultTimeout,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			receivedDuration := GetTimeout(tc.inputDuration)
+			if receivedDuration != tc.expectedDuration {
+				t.Errorf("expected %q received %q", tc.expectedDuration.String(), receivedDuration.String())
+			}
+		})
+	}
+}
+
+// TestSetTaskRunTimer checks that the SetTaskRunTimer method correctly calls the TaskRun
+// callback after a set amount of time.
+func TestSetTaskRunTimer(t *testing.T) {
+	taskRun := tb.TaskRun("test-taskrun-arbitrary-timer", testNs, tb.TaskRunSpec(
+		tb.TaskRunTaskRef(simpleTask.Name, tb.TaskRefAPIVersion("a1")),
+		tb.TaskRunTimeout(2*time.Second),
+	), tb.TaskRunStatus(tb.Condition(apis.Condition{
+		Type:   apis.ConditionSucceeded,
+		Status: corev1.ConditionUnknown}),
+		tb.TaskRunStartTime(time.Now().Add(-10*time.Second)),
+	))
+
+	d := test.Data{
+		TaskRuns: []*v1alpha1.TaskRun{taskRun},
+		Tasks:    []*v1alpha1.Task{simpleTask},
+		Namespaces: []*corev1.Namespace{{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testNs,
+			},
+		}},
+	}
+	stopCh := make(chan struct{})
+	c, _ := test.SeedTestData(t, d)
+	observer, _ := observer.New(zap.InfoLevel)
+	testHandler := NewTimeoutHandler(c.Kube, c.Pipeline, stopCh, zap.New(observer).Sugar())
+	timerDuration := 50 * time.Millisecond
+	timerFailDeadline := 100 * time.Millisecond
+	doneCh := make(chan struct{})
+	callback := func(_ interface{}) {
+		close(doneCh)
+	}
+	testHandler.SetTaskRunCallbackFunc(callback)
+	go testHandler.SetTaskRunTimer(taskRun, timerDuration)
+	select {
+	case <-doneCh:
+		// The task run timer executed before the failure deadline
+	case <-time.After(timerFailDeadline):
+		t.Errorf("timer did not execute task run callback func within expected time")
+	}
+}
+
+// TestBackoffDuration asserts that the backoffDuration func returns Durations
+// within the timeout handler's bounds.
+func TestBackoffDuration(t *testing.T) {
+	testcases := []struct {
+		description      string
+		inputCount       uint
+		jitterFunc       func(int) int
+		expectedDuration time.Duration
+	}{
+		{
+			description:      "an input count that is too large is rounded to the maximum allowed exponent",
+			inputCount:       uint(maxBackoffExponent + 1),
+			jitterFunc:       func(in int) int { return in },
+			expectedDuration: maxBackoffSeconds * time.Second,
+		},
+		{
+			description:      "a jittered number of seconds that is above the maximum allowed is constrained",
+			inputCount:       1,
+			jitterFunc:       func(in int) int { return maxBackoffSeconds + 1 },
+			expectedDuration: maxBackoffSeconds * time.Second,
+		},
+	}
+	packageJitterFunc := jitterFunc
+	for _, tc := range testcases {
+		t.Run(tc.description, func(t *testing.T) {
+			jitterFunc = tc.jitterFunc
+			result := backoffDuration(tc.inputCount)
```

Comment:
I wonder if we could pass the `jitterFunc as an argument to `backoffDuration` instead. This would remove the need to set (or have) the `jitterFunc` global variable (just for the test) :angel:

Reply:
```go
# in code
backoffDuration(b.NumAttempts, rand.Intn)
# in tests
backoffDuration(tc.inputCount, tc.jitterFunc)
```